packages/workflow/src/workflow.ts

```diff
@@ -1356,21 +1384,119 @@ export function upsertSearchAttributes(searchAttributes: SearchAttributes): void
     throw new Error('searchAttributes must be a non-null SearchAttributes');
   }
 
-  activator.pushCommand({
-    upsertWorkflowSearchAttributes: {
-      searchAttributes: mapToPayloads(searchAttributePayloadConverter, searchAttributes),
-    },
-  });
+  if (Array.isArray(searchAttributes)) {
+    // Typed search attributes
+    activator.pushCommand({
+      upsertWorkflowSearchAttributes: {
+        searchAttributes: mapToPayloads<string, TypedSearchAttributeUpdateValue<SearchAttributeType>>(
```

Comment:
Why are you using `mapToPayloads` here rather than `encodeSearchAttributes()`? If needed, you can make `encodeSearchAttributes()` allow undefined on either argument so that it is usable for this purpose.

Same for legacy SA below.